segfault TyWithLocation, `extern C`

<!--
File: /tmp/im3/6E002F7CAF4B514AABA22496E8A0D08DB743C02812EAA8D5F63CDCB32BBC3B9B.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -frust-incomplete-and-experimental-compiler-do-not-use
trait MemoryUnit {
    extern "C" fn read_dword(&'s self) -> u16 {}
}

impl MemoryUnit for MemoryUnit {
    extern "C" fn read_dword(&'s self) -> u16 {
        let b16 = self.read_word() as u16;

        b16 << 8
    }
}

````

original:
````rust
trait MemoryUnit {
    extern "C" fn read_dword(&'s self) -> u16 {
        let a16 = self.read_word() as u16;
        let b16 = self.read_word() as u16;

        (b16 << 8) | a16
    }
}

struct ROM {}

impl MemoryUnit for MemoryUnit {
    extern "C" fn read_dword(&'s self) -> u16 {
        let a16 = self.read_word() as u16;
        let b16 = self.read_word() as u16;

        b16 << 8
    }
}

````

Version information:

5a762b9243e834d69f90985950f4ef330f16b34f



Command:
`/home/matthias/vcs/github/gccrs/gccrs-build/gcc/crab1 -frust-incomplete-and-experimental-compiler-do-not-use`

<details><summary><strong>Program output</strong></summary>
<p>

```
/tmp/icemaker_global_tempdir.CmTzlVK6SkSa/rustc_testrunner_tmpdir_reporting.eNpDdPpl5c3x/mvce.rs:2:29: error: failed to resolve lifetime
    2 |     extern "C" fn read_dword(&'s self) -> u16 {}
      |                             ^
/tmp/icemaker_global_tempdir.CmTzlVK6SkSa/rustc_testrunner_tmpdir_reporting.eNpDdPpl5c3x/mvce.rs:6:29: note: failed to resolve lifetime
    6 |     extern "C" fn read_dword(&'s self) -> u16 {
      |                             ^
/tmp/icemaker_global_tempdir.CmTzlVK6SkSa/rustc_testrunner_tmpdir_reporting.eNpDdPpl5c3x/mvce.rs:6:29: note: failed to resolve lifetime
/tmp/icemaker_global_tempdir.CmTzlVK6SkSa/rustc_testrunner_tmpdir_reporting.eNpDdPpl5c3x/mvce.rs:2:29: error: failed to resolve lifetime
    2 |     extern "C" fn read_dword(&'s self) -> u16 {}
      |                             ^
crab1: internal compiler error: Segmentation fault
0x301f909 internal_error(char const*, ...)
	../../gcc/diagnostic-global-context.cc:786
0x199fb9f crash_signal
	../../gcc/toplev.cc:323
0xfe8180 Rust::TyTy::BaseType::get_ref() const
	../../gcc/rust/typecheck/rust-tyty.cc:168
0x100002c Rust::TyTy::TyWithLocation::TyWithLocation(Rust::TyTy::BaseType*)
	../../gcc/rust/typecheck/rust-tyty-util.cc:131
0x107c7da Rust::Resolver::TypeCheckImplItemWithTrait::visit(Rust::HIR::Function&)
	../../gcc/rust/typecheck/rust-hir-type-check-implitem.cc:592
0x107b97e Rust::Resolver::TypeCheckImplItemWithTrait::Resolve(Rust::HIR::ImplBlock&, Rust::HIR::ImplItem&, Rust::TyTy::BaseType*, Rust::TyTy::TypeBoundPredicate&, std::vector<Rust::TyTy::SubstitutionParamMapping, std::allocator<Rust::TyTy::SubstitutionParamMapping> >)
	../../gcc/rust/typecheck/rust-hir-type-check-implitem.cc:439
0x103f524 Rust::Resolver::TypeCheckItem::validate_trait_impl_block(Rust::Resolver::TraitReference*, Rust::HIR::ImplBlock&, Rust::TyTy::BaseType*, std::vector<Rust::TyTy::SubstitutionParamMapping, std::allocator<Rust::TyTy::SubstitutionParamMapping> >&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:795
0x1044e4d Rust::Resolver::TypeCheckItem::visit(Rust::HIR::ImplBlock&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:493
0x103ac05 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:57
0xfda3bb Rust::Resolver::TypeResolution::Resolve(Rust::HIR::Crate&)
	../../gcc/rust/typecheck/rust-hir-type-check.cc:71
0xe875ec Rust::Session::compile_crate(char const*)
	../../gcc/rust/rust-session-manager.cc:682
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.

```

</p>
</details>

--- a/gcc/rust/typecheck/rust-hir-type-check-implitem.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-implitem.cc
@@ -558,6 +558,8 @@ TypeCheckImplItemWithTrait::visit (HIR::Function &function)
   // normal resolution of the item
   TyTy::BaseType *lookup
     = TypeCheckImplItem::Resolve (parent, function, self, substitutions);
+  if (lookup == nullptr)
+    return;
 
   // map the impl item to the associated trait item
   const auto tref = trait_reference.get ();

--- a/gcc/rust/typecheck/rust-hir-type-check.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check.cc
@@ -237,19 +237,20 @@ TraitItemReference::get_type_from_fn (/*const*/ HIR::TraitItemFunc &fn) const
 		      : Mutability::Mut;
 		rust_assert (self_param.has_lifetime ());
 
+		auto region = TyTy::Region::make_anonymous ();
 		auto maybe_region = context->lookup_and_resolve_lifetime (
 		  self_param.get_lifetime ());
-
-		if (!maybe_region.has_value ())
+		if (maybe_region.has_value ())
+		  region = maybe_region.value ();
+		else
 		  {
 		    rust_error_at (self_param.get_locus (),
 				   "failed to resolve lifetime");
-		    return get_error ();
 		  }
+
 		self_type = new TyTy::ReferenceType (
 		  self_param.get_mappings ().get_hirid (),
-		  TyTy::TyVar (self->get_ref ()), mutability,
-		  maybe_region.value ());
+		  TyTy::TyVar (self->get_ref ()), mutability, region);
 	      }
 	      break;
 

--- a/gcc/testsuite/rust/compile/issue-4188.rs
+++ b/gcc/testsuite/rust/compile/issue-4188.rs
@@ -1,0 +1,13 @@
+trait MemoryUnit {
+    extern "C" fn read_dword(&'s self) -> u16 {}
+    // { dg-error {failed to resolve lifetime} "" { target *-*-* } .-1 }
+    // { dg-error {mismatched types} "" { target *-*-* } .-2 }
+}
+
+impl MemoryUnit for MemoryUnit {
+    extern "C" fn read_dword(&'s self) -> u16 {
+        let b16 = self.read_word() as u16;
+
+        b16 << 8
+    }
+}

